### PR TITLE
core: physics: compute the global energy of an envelope

### LINKED ISSUE #2597: Display the total energy consumption of the train
## Description

Compute the energy consumed by the train along its journey, and display the total energy in front-end. This value of energy changes when applying allowances, so the different values depending on the allowance must be displayed.

## Goal

Prove that OSRD will help saving energy thanks to MARECO algorithm.

## Acceptance criteria

- [ ] The values must be displayed under the origin and destination of the train

Image

- [ ] The values must be displayed in kWh, rounded up to the unit
- [ ] The text introducing the value must be the following : "Energie consommée à la jante"
- [ ] Two values must be displayed : one for the basic run (_marche de base_) and one with the allowance if it is computed (_avec marge de régularité_)

### PR DESCRIPTION
**Pull request by axrolld**
Global computation. Not the same as the attempt done in #2600 .
Closes #2597 .

### CODE REVIEW

File: api/openapi.yaml [hunk 1]
```diff
@@ -908,6 +908,8 @@ components:
                 items: { type: number, format: float }
                 minItems: 2
                 maxItems: 2
+        energy_consumed:
```
Comment (multun) [inline api/openapi.yaml]: I really don't like the name: `mechanical_effort` would be more fitting. "Energy consumption" somewhat implies this number takes into account system efficiency (like engine efficiency and power line losses), which it does not

File: core/src/main/java/fr/sncf/osrd/envelope/Envelope.java [hunk 2]
```diff
@@ -117,6 +120,18 @@ public double getMinSpeed() {
...
+    public double getEnergyConsumed(
+            PhysicsPath path,
+            PhysicsRollingStock rollingStock
+    ) {
+        var cumulativeEnergy = 0.0;
+        for (var part : parts) {
+            cumulativeEnergy += part.getEnergyConsumed(path, rollingStock);
+        }
+        return cumulativeEnergy;
+    }
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/envelope/Envelope.java]: This isn't the right place to add this helper: the `envelope` folder is a separate module from the `envelope_sim` module. Adding it as a static function somewhere else helps avoid a circular dependency. It does not matter now, but will when we will make the app more modular.

File: core/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.java [hunk 3]
```diff
@@ -333,6 +343,45 @@ public long getTotalTimeMS(int pointIndex) {
         return getTotalTimesMS()[pointIndex];
     }
 
+    /** Returns the total energy consumed at train wheel on this envelopePart */
+    public double getEnergyConsumed(
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.java]: same as above, I know it's less convenient, but this stuff should really be kept out of the `envelope` module, as it does not have to be here: it can just use the public API

File: core/src/test/java/fr/sncf/osrd/train/TestTrains.java [hunk 4]
```diff
@@ -3,52 +3,55 @@
...
+    public static final RollingStock CONSTANT_POWER_TRAIN;
 
-    static {
-        double trainMass = 850000; // in kilos
-        double maxSpeed = 300 / 3.6;
-        var tractiveEffortCurve = new ArrayList<RollingStock.TractiveEffortPoint>();
+    private enum CurveShape {
+        LINEAR,
+        HYPERBOLIC
+    }
```
Comment (multun) [inline core/src/test/java/fr/sncf/osrd/train/TestTrains.java:23]: I'm pretty confused as to why this change was introduced in this pull request, can you elaborate and add a source comment explaining what's the difference between linear and hyperbolic curves, and what it's used for?

**Review by multun — CHANGES_REQUESTED** [review]
Thanks for your contribution :+1:  the code looks good overall, A few things don't quite blend in but I'm sure these can be ironed out quickly

File: api/openapi.yaml (hunk 1 above)
Comment (axrolld) [reply api/openapi.yaml]: I'd rather name it `mechanical_energy_consumed`, as this this an energy and not an effort. But you're right it's good to precise that it does not take into account the entire energy channel.

File: core/src/main/java/fr/sncf/osrd/envelope/Envelope.java (hunk 2 above)
Comment (axrolld) [reply core/src/main/java/fr/sncf/osrd/envelope/Envelope.java]: You're right, that sounds like a better way to implement this. We'll change it.

File: core/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.java (hunk 3 above)
Comment (axrolld) [reply core/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.java]: Done. We moved it to `EnvelopePhysics.java`

File: core/src/main/java/fr/sncf/osrd/envelope/Envelope.java (hunk 2 above)
Comment (axrolld) [reply core/src/main/java/fr/sncf/osrd/envelope/Envelope.java]: Moved to `EnvelopePhysics.java`.

File: core/src/test/java/fr/sncf/osrd/train/TestTrains.java (hunk 4 above)
Comment (axrolld) [reply core/src/test/java/fr/sncf/osrd/train/TestTrains.java:23]: Alright we'll add some comments.

File: core/src/main/java/fr/sncf/osrd/envelope/EnvelopePhysics.java [hunk 5]
```diff
@@ -174,4 +179,59 @@ public static double intersectStepWithSpeed(
...
+            var nextPos = part.getPointPos(i + 1);
+            var positionDelta = nextPos - pos;
+            workDrag -= rollingStock.getRollingResistance(meanSpeed) * positionDelta;
+        }
+
+        var totalEnergyConsumed = kineticEnergyDelta - workGravity - workDrag;
+
+        // If the train is braking the result should be negative
+        if (part.hasAttr(EnvelopeProfile.BRAKING))
+            assert totalEnergyConsumed <= 0;
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/envelope/EnvelopePhysics.java]: isn't this assert not always true? I call we encountered such a case while testing. it should also be 
```suggestion
assert !part.hasAttr(EnvelopeProfile.BRAKING) || totalEnergyConsumed <= 0;
```

File: core/src/main/java/fr/sncf/osrd/standalone_sim/result/ResultTrain.java [hunk 6]
```diff
@@ -16,19 +16,23 @@ public class ResultTrain {
     public final Map<String, ResultOccupancyTiming> routeOccupancies;
     @Json(name = "signal_updates")
     public final Collection<SignalUpdate> signalUpdates;
+    @Json(name = "mechanical_energy_consumed")
+    public final double energyConsumed;
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/standalone_sim/result/ResultTrain.java]: please sync the field name with the json name

File: core/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.java [hunk 7]
```diff
@@ -481,7 +487,7 @@ public Double interpolatePosition(double speed) {
     /** Check if a is in the interval [b, c] or [c, b]*/
     private static boolean isBetween(double a, double b, double c) {
         var min = Math.min(b, c);
-        var max = Math.max(b, c);
+        var max = max(b, c);
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.java]: that's probably not intended

File: core/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.java [hunk 8]
```diff
@@ -1,6 +1,8 @@
 package fr.sncf.osrd.envelope.part;
 
 import static fr.sncf.osrd.envelope.EnvelopePhysics.intersectStepWithSpeed;
+import static java.lang.Math.abs;
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.java]: the import is not used

File: core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.java [hunk 9]
```diff
@@ -56,12 +58,17 @@ public static ResultTrain run(
         // Compute events
         var events = computeEvents(infra, trainPath, trainLength, envelopeWithStops);
 
+        // Compute energy consumed
+        var envelopePath = EnvelopeTrainPath.from(trainPath);
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.java:62]: the envelopetrainpath was already computed before, can't it be passed as a parameter?

**Review by multun — CHANGES_REQUESTED** [review]

**Review by bgiuliana — DISMISSED** [review]

File: core/src/main/java/fr/sncf/osrd/envelope/EnvelopePhysics.java (hunk 5 above)
Comment (axrolld) [reply core/src/main/java/fr/sncf/osrd/envelope/EnvelopePhysics.java]: It's supposed to be, we just assert this for debugging purposes. Your suggestion makes sense though

File: core/src/main/java/fr/sncf/osrd/standalone_sim/result/ResultTrain.java (hunk 6 above)
Comment (axrolld) [reply core/src/main/java/fr/sncf/osrd/standalone_sim/result/ResultTrain.java]: Ah yeah you're right

File: core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.java (hunk 9 above)
Comment (axrolld) [reply core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.java:62]: We need an EnvelopePath for the method `getMechanicalEnergyConsumed`, and `trainPath` is a TrainPath, so we need to convert it. We can't find the `envelopetrainpath` you're mentionning above :/

Comment (eckter) [reply core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.java:62]: When it's called from `StandaloneSim` we have an `EnvelopeTrainPath` available, but we also call it from STDCM, and it's not readily available there

Comment (axrolld) [reply core/src/main/java/fr/sncf/osrd/standalone_sim/ScheduleMetadataExtractor.java:62]: In STDCM context, no `envelopePath` is computed before calling the `run` method from `ScheduleMetaDataExtractor`, so it would be quite a pain to modify everything.

File: core/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.java (hunk 8 above)
Comment (axrolld) [reply core/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.java]: Removed, thank you !

File: core/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.java (hunk 7 above)
Comment (axrolld) [reply core/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.java]: Indeed, oops

**Review by multun — APPROVED** [review]
